### ISSUE TITLE
Qwen2.5-Coder-32B ci-eval-32 greedy non-determinism: residual source is NOT the gather race (investigation + next steps)

### ISSUE THREAD

**Issue by mtairum** [tenstorrent/tt-metal#48614]
## Summary
`Qwen2.5-Coder-32B` fails the `ci-eval-32` **repeat-batch consistency check** on `wh_llmbox_perf` (T3000, 8-device): the same prompt run at a different user slot across batches produces different tokens (1–2 of 64 checks fail, intermittently). This is a **greedy-decode** non-determinism whose source is **NOT** the candidate all-gather race fixed in #48404, and **NOT** stochastic-sampling RNG. Filing as a consolidated investigation with next steps.

Pre-existing / related trackers: #47823, #43626, #38645 (same failure). Root-cause context for the gather race: #48222, #48469, fixed in #48404.

## What the test actually checks
- Test: `models/tt_transformers/demo/simple_text_demo.py::test_demo_text[...-performance-ci-eval-32]`
- `ci-eval-32` config: `repeat_batches=3`, `batch_size=32`, **shifting prompts**, `sampling_params = {temperature: 0, top_p: 0.08, top_k: 32}` → **temperature=0 ⇒ greedy/argmax** (the generator rewrites temp=0 to `k=1, p=0, temp=1`).
- The consistency check compares the **same prompt at a different user slot** across consecutive batches and asserts identical output (`simple_text_demo.py` ~L1649).

## Key findings
1. **It's greedy, so there is no RNG in this path.** The per-user seed / `rand_tile_init(seed==0)` behaviour does not apply — argmax does not consume the random value. So the mismatches are a genuine non-deterministic *computation*, not sampler RNG variance.
2. **Not the candidate-gather race (#48222/#48469).** Baselines:
   - main (broken gather): **1/64** failing (per #47823).
   - this branch with the #48404 barrier fix: **2/64** failing.
   - i.e. the gather fix made **no material difference** for Coder (1 vs 2 of 64 is noise for an intermittent bug). By contrast it **fully cleared Mixtral-8x7B** (#48356: 3–4/64 → 0/64). Same test, same params — so Coder's source is model-specific and *not* the gather.
3. **The divergence is one borderline prompt.** Both failing checks in run [28399871993](https://github.com/tenstorrent/tt-metal/actions/runs/28399871993) (job 84155934408) were `prompt[1]` at different slots:
   - `User 1 (batch 0)` vs `User 0 (batch 1)` — both prompt[1] — differ
   - `User 0 (batch 1)` vs `User 31 (batch 2)` — both prompt[1] — differ
   So a specific prompt's greedy output is not perfectly **slot-invariant**.

## Leading hypothesis
**Slot-dependent floating-point rounding in the batch-32 reductions flips a near-tie argmax.** Batch reductions (attention/MLP, and the topk/gather reduction) are not associative, so the same prompt at a different slot can get logits differing by < bf16 epsilon, flipping top-1 where two candidates are near-equal. Coder's large vocab (152064 vs Mixtral's 32000) gives flatter, more tie-prone distributions → more borderline positions. Consistent with #47823's "reductions/ordering" suspect.

Ruled out: stochastic RNG (greedy); the candidate-gather barrier (#48404, neutral for Coder).
Not yet ruled out: exact bf16 ties vs near-ties; a slot-dependent layout/padding effect; KV-cache ordering.

## Next steps (instrumentation plan — same technique used for #48222)
1. Repro and capture which token position first diverges for `prompt[1]` between the two slots.
2. At that decode step, dump (via `ttnn.get_device_tensors` on host, non-perturbing per-device shards) the **gathered candidate values + indices** for that user at both slots, and the **full pre-topk logits**, and compare:
   - If the gathered candidate **values are identical** across slots but the **argmax index differs** → exact bf16 tie in the op's tie-break.
   - If the candidate **values differ by ~bf16 epsilon** across slots → upstream reduction rounding (attention/MLP), i.e. logits aren't slot-invariant.
   - If values differ materially → a real per-slot compute bug (KV-cache/page-table/mask).
3. Localize the first non-slot-invariant op by comparing per-op activations for the same prompt at slot i vs slot j (forward only, batch-32), bisecting attention → MLP → norm → lm_head → topk → gather.
4. If it's bf16 ties: consider a deterministic tie-break in the topk/argmax (e.g. lowest global index wins) or fp32 accumulation at the reduction; if it's reduction rounding: evaluate fp32 accumulation / fixed reduction order for the affected op.

## Repro
On a `wh_llmbox_perf` (T3000, 8-device) runner:
```bash
export HF_MODEL=Qwen/Qwen2.5-Coder-32B-Instruct TT_CACHE_PATH=/mnt/MLPerf/huggingface/tt_cache/Qwen/Qwen2.5-Coder-32B-Instruct
pytest --timeout 1800 models/tt_transformers/demo/simple_text_demo.py -k "performance-ci-eval-32"
```
Failing example (this branch, with #48404): https://github.com/tenstorrent/tt-metal/actions/runs/28399871993/job/84155934408 — "2 out of 64 consistency checks failed".

## Scope note
This is **not** introduced by #48404 (pre-existing on main) and is a different bug class than the gather race. It should not block #48404.
